### PR TITLE
Add per ip rate limiter

### PR DESCRIPTION
**Pull request by yilunzhang**
Signed-off-by: Yilun <zyl.skysniper@gmail.com>

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### CODE REVIEW

File: node/localnode.go [hunk 1]
```diff
@@ -110,6 +112,18 @@ func NewLocalNode(wallet *vault.Wallet, nn *nnet.NNet) (*LocalNode, error) {
 	event.Queue.Subscribe(event.BlockPersistCompleted, localNode.cleanupTransactions)
 	event.Queue.Subscribe(event.NewBlockProduced, localNode.CheckIDChange)
 
+	nn.MustApplyMiddleware(nnetnode.ConnectionAccepted{func(conn net.Conn) (bool, bool) {
+		host, _, err := net.SplitHostPort(conn.RemoteAddr().String())
```
Comment (bufrr) [inline node/localnode.go:116]: `host` including IP and port, maybe we need to limit the IP access rate without port

Comment (yilunzhang) [reply node/localnode.go:116]: host does not include port, check https://pkg.go.dev/net#SplitHostPort

File: api/websocket/server/server.go [hunk 2]
```diff
@@ -247,6 +248,16 @@ func (ws *WsServer) Stop() {
 
 //websocketHandler
 func (ws *WsServer) websocketHandler(w http.ResponseWriter, r *http.Request) {
+	host, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err == nil {
```
Comment (bufrr) [inline api/websocket/server/server.go:252]: Do we need to check if `err != nil` ?

Comment (yilunzhang) [reply api/websocket/server/server.go:252]: In theory it shouldn't happen, and even it does, we can't do anything about it except for print it...

File: api/httpjson/RPCserver.go [hunk 3]
```diff
@@ -59,7 +60,18 @@ func NewServer(localNode *node.LocalNode, wallet *vault.Wallet) *RPCServer {
...
+		limiter := ratelimiter.GetLimiter("rpc:"+host, config.Parameters.RPCIPRateLimit, int(config.Parameters.RPCIPRateBurst))
+		if !limiter.Allow() {
+			log.Infof("RPC connection limit of %s reached", host)
+			w.WriteHeader(http.StatusTooManyRequests)
+			return
+		}
+	}
+
 	if !s.limiter.Allow() {
+		log.Infof("RPC connection limit reached", host)
```
Comment (bufrr) [inline api/httpjson/RPCserver.go:74]: `host` havn't added to the log string

Comment (yilunzhang) [reply api/httpjson/RPCserver.go]: Fixed, no need to log host because this is overall rate limit

**Review by bufrr — APPROVED** [review]

File: api/ratelimiter/ratelimiter.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+	if limiter, ok := rateLimiters[key]; ok {
+		return limiter
+	}
+
+	limiter := rate.NewLimiter(rate.Limit(limit), burst)
+	rateLimiters[key] = limiter
+
+	return limiter
+}
```
Comment (gdmmx) [inline api/ratelimiter/ratelimiter.go:24]: Which timing to remove key/Limiter obj from Map even conn/request had finished?

**Review by gdmmx — APPROVED** [review]
LGTM

**Pull request merged**